packages/@aws-cdk/core/lib/validation/private/report.ts

```diff
@@ -0,0 +1,222 @@
+        ...Object.entries(plugin.summary.metadata ?? {}),
+      ]));
+
+      if (plugin.violations) {
+        output.push('');
+        output.push('(Violations)');
+      }
+
+      plugin.violations.forEach((violation) => {
+        const constructs = violation.violatingConstructs;
+        const occurrences = constructs.length;
+        const title = reset(red(bright(`${violation.ruleName} (${occurrences} occurrences)`)));
+        output.push('');
+        output.push(title);
+        output.push('');
+        output.push('  Occurrences:');
+        for (const construct of constructs) {
+          output.push('');
+          output.push(`    - Construct Path: ${construct.constructPath}`);
+          output.push(`    - Template Path: ${construct.templatePath}`);
+          output.push(`    - Creation Stack:\n\t${construct.constructStack}`);
+          output.push(`    - Resource Name: ${construct.resourceName}`);
+          if (construct.locations) {
+            output.push('    - Locations:');
+            for (const location of construct.locations) {
+              output.push(`      > ${location}`);
+            }
+          }
+        }
+        output.push('');
+        output.push(`  Recommendation: ${plugin.violations[0].recommendation}`);
+        if (plugin.violations[0].fix) {
+          output.push(`  How to fix: ${plugin.violations[0].fix}`);
+        }
+      });
+
+    });
+
+    return output.join(os.EOL);
+  }
+
+  /**
+   * Get the stack trace from the construct node metadata.
+   * The stack trace only gets recorded if the node is a `CfnResource`,
+   * but the stack trace will have entries for all types of parent construct
+   * scopes
+   */
+  private getTraceMetadata(node?: Node): string[] {
+    if (node) {
+      const construct = this.tree.getConstructByPath(node.path);
+      if (construct) {
+        const trace = construct.node.defaultChild?.node.metadata.find(meta => !!meta.trace)?.trace ?? [];
+        return Object.create(trace);
+      }
+    }
+    return [];
+  }
+
+  /**
+   * Construct the stack trace of constructs. This will start with the
+   * resource that has a violation and then go up through it's parents
+   */
+  private getConstructTrace(node: Node, locations?: string[]): ConstructTrace {
+    const trace = this.tree.getTraceFromCache(node.path);
+    if (trace) {
+      return trace;
+    }
+    const metadata = locations ?? this.getTraceMetadata(node);
+    const thisLocation = metadata.shift();
+    const constructTrace = {
+      id: node.id,
+      path: node.path,
+      parent: node.parent ? this.getConstructTrace(node?.parent, metadata) : undefined,
+      library: node.constructInfo?.fqn,
+      libraryVersion: node.constructInfo?.version,
+      location: thisLocation,
+    };
+    this.tree.setTraceCache(constructTrace.path, constructTrace);
+    return constructTrace;
+  }
+
+  /**
+   * - Creation Stack:
+   *     └──  Bucket (validator-test/MyCustomL3Construct/Bucket)
```

Comment:
Question about the order of the stack. Should we start with the resource that has the violation on top and then go up the tree? Or should we start with the top of the tree and go down, stopping at the violation?